patches/sitecustomize.py

```diff
@@ -1,22 +1,56 @@
 # TODO(rosbo): Remove this once we fix the issue with fastai importing older libcudnn if imported prior to tensorflow
 import tensorflow
 import os
+from google.auth import credentials
+from google.cloud import bigquery
+from google.cloud.bigquery._http import Connection
 
-# Monkey patches BigQuery client creation to use proxy.
-kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
-if kaggle_proxy_token:
-    from google.auth import credentials
-    from google.cloud import bigquery
-    from google.cloud.bigquery._http import Connection
+class KaggleKernelCredentials(credentials.Credentials):
+    def refresh(self, request):
+        print("Calling Kaggle.UserSecrets to refresh token.")
+        # Set self.token and self.expiry here.
+        raise NotImplementedError("Private BigQuery integration is not yet implemented.")
 
-    Connection.API_BASE_URL = os.getenv("KAGGLE_DATA_PROXY_URL")
-    Connection._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = kaggle_proxy_token
+kaggle_proxy_token = os.getenv("KAGGLE_DATA_PROXY_TOKEN")
+CONNECTION_BASE_URL = Connection.API_BASE_URL
+def monkeypatch_bq(bq_client, *args, **kwargs):
+    data_proxy_project = os.getenv("KAGGLE_DATA_PROXY_PROJECT")
+    bq_user_jwt = os.getenv("KAGGLE_BQ_USER_JWT")
+    specified_project = kwargs.get('project').lower() if 'project' in kwargs else None
+    # Use Data Proxy if user has specified to use the Kaggle project, or if
+    # there are no connected GCP accounts (to maintain backwards compatibility).
+    if bq_user_jwt is None and specified_project and specified_project != 'kaggle':
+        raise Exception("In order to query a private BigQuery project, please connect a GCP account. "
+                        "Otherwise specify 'kaggle' as the project to use Kaggle's public dataset BigQuery integration.")
+    use_data_proxy = specified_project == 'kaggle' or bq_user_jwt is None
+    if use_data_proxy:
+        if data_proxy_project is None or kaggle_proxy_token is None:
+            # We don't have the data proxy info so leave the bq client unmodified.
+            return bq_client(*args, **kwargs)
+        print("Using Kaggle's public dataset BigQuery integration.")
+        Connection.API_BASE_URL = os.getenv("KAGGLE_DATA_PROXY_URL")
+        Connection._EXTRA_HEADERS["X-KAGGLE-PROXY-DATA"] = kaggle_proxy_token
+        anon_credentials = credentials.AnonymousCredentials()
+        anon_credentials.refresh = lambda *args: None
+        kwargs['project'] = data_proxy_project
+        return bq_client(
+            *args,
+            credentials=anon_credentials,
+            **kwargs)
+    else:
+        Connection.API_BASE_URL = CONNECTION_BASE_URL
```

Comment:
Not sure why you need this and `CONNECTION_BASE_URL` at all.  Why not just leave it to what it is in the first place?

Reply:
My thinking was that, the first time this runs on startup, we store `CONNECTION_BASE_URL = Connection.API_BASE_URL` as the default url that the bigquery library wants to use. And then each time the user uses the bigquery client library, we will update Connection.API_BASE_URL to either the data proxy url (line 32) or back to the default url (line 41) depending on what project the user uses (Ie. if user indicates they want to use the kaggle proxy or their own account). 

Setting it this way, is only to prevent issues if the user first instantiates the client for the kaggle proxy and then switches back to using their own project or vice versa.

Reply:
> Setting it this way, is only to prevent issues if the user first instantiates the client for the kaggle proxy and then 
> switches back to using their own project or vice versa.

Can we actually support that?  This is a static class variable we're changing here right?  Is an existing client guaranteed not to reuse this?

Reply:
That's a good point. This will work if the user creates clientA (kaggle proxy) and uses it then creates clientB (non proxy) and uses it, but won't work if they interleave their usage. It's a little less broken but still broken - do you have any ideas on how to get this to work or should we disallow?

One way I can imagine us disallowing it is checking if the current client making a call is of the same type as the connection (ie. a project=kaggle client call should ensure that the Connection is the one expected otherwise throw an error). Now that I think of it, we can probably intercept each client call and update the Connection.API_BASE_URL at the time of the call so that it's always the correct value. How does that sound?

Reply:
The right solution is not to do something that isn't supported :).  The existing approach is already a hack since it accesses internals and modifies a static that I don't believe is meant to be changed.

So, the right solution is to modify upstream to support this all properly.  IIRC this was my original suggestion already.  Monkey patching is only at most okay in the short term.